plugin/evm/ftso_system_test.go

```diff
@@ -0,0 +1,221 @@
+	testAbiSubmitter, _  = abi.JSON(strings.NewReader(testAbiSubmitterJSON))
+)
+
+func TestFTSOSystem_Contracts(t *testing.T) {
+	t.Run("common case", func(t *testing.T) {
+		auth, be := simulatedBlockchain(t)
+
+		// VotePower contract
+		vpInt := big.NewInt(100)
+		votePowerAddr := deployTestContract(t, auth, be, testAbiVotePower, testAbiVotePowerBin, vpInt)
+
+		// Wnat contract
+		wnatAddr := deployTestContract(t, auth, be, testAbiWnat, testAbiWnatBin, votePowerAddr)
+
+		// Reward contract
+		unclaimedReward := big.NewInt(100)
+		rewardAddr := deployTestContract(t, auth, be, testAbiReward, testAbiRewardBin, wnatAddr, unclaimedReward)
+
+		// Manager contract
+		rewardEpochDurationSeconds := big.NewInt(100)
+		rewardEpochPowerHeight := big.NewInt(100)
+		rewardEpochStartHeight := big.NewInt(100)
+		rewardEpochStartTime := big.NewInt(100)
+		rewardEpochsStartTs := big.NewInt(1) // this should be less than header.Time()
+		ftsoManagerAddr := deployTestContract(t, auth, be, testAbiManager, testAbiManagerBin, rewardAddr, rewardEpochDurationSeconds, rewardEpochPowerHeight, rewardEpochStartHeight, rewardEpochStartTime, rewardEpochsStartTs)
+
+		// Registry contract
+		index1 := big.NewInt(123)
+		index2 := big.NewInt(321)
+		supportedIndices := []*big.Int{index1, index2}
+		ftsoRegistryAddr := deployTestContract(t, auth, be, testAbiFtsoRegistry, testAbiFtsoRegistryBin, supportedIndices)
+
+		// Whitelist contract
+		indices := []*big.Int{big.NewInt(123)}
+		priceProvidersAddresses := []common.Address{common.BigToAddress(big.NewInt(222))}
+		voterWhitelisterAddr := deployTestContract(t, auth, be, testAbiVoterWhitelister, testAbiVoterWhitelisterBin, indices, priceProvidersAddresses)
+
+		// Validation contract
+		providerAddresses := []common.Address{common.BigToAddress(big.NewInt(222))}
```

Comment:
Again, all values should be unique. It's a good idea to actually define all values at the very top of the test, and then use named variables in the initialization logic.